libraries/app/database_api.cpp

```diff
@@ -1149,6 +1154,64 @@ void database_api_impl::unsubscribe_from_market(asset_id_type a, asset_id_type b
    _market_subscriptions.erase(std::make_pair(a,b));
 }
 
+string database_api_impl::uint128_amount_to_string( const fc::uint128& amount, const uint8_t precision )const
+{ try {
+   fc::uint128 scaled_precision = 1;
+   for( uint8_t i = 0; i < precision; ++i )
+      scaled_precision *= 10;
+   FC_ASSERT(scaled_precision > 0);
+
+   string result = string( amount / scaled_precision );
+   auto decimals = amount % scaled_precision;
+   if( decimals > 0 )
+      result += "." + string(scaled_precision + decimals).erase(0,1);
+   return result;
+} FC_CAPTURE_AND_RETHROW( (amount)(precision) ) }
+
+string database_api_impl::price_to_string( const price& _price, const asset_object& _base, const asset_object& _quote )const
+{ try {
+   share_type base_amount = _price.base.amount;
+   share_type quote_amount = _price.quote.amount;
+   if( _price.base.asset_id != _base.id )
+      std::swap( base_amount, quote_amount );
+
+   // times (10**18 * 10**3) so won't overflow but always have good accuracy
+   fc::uint128 price128 = fc::uint128( base_amount.value ) * uint64_t(1000000000000000000ll) * uint64_t(1000) / quote_amount.value;
```

Comment:
I think this can overflow

Reply:
Since `GRAPHENE_MAX_SHARE_SUPPLY` is defined as `10^15`, in practice a price should be in the range of `[1/10^15, 10^15/1]`, based on this assumption, it won't overflow here because the biggest `uint128` number is greater than `10^38`. However,
* there are edge cases that `base` or `quote` can be `0`, which should be caught somewhere;
* actually there is another issue: the order matching engine assumes all prices are in the range mentioned above, but the code didn't enforce the price to be in the rage when a new order is created, perhaps need a hard fork to fix.

Reply:
Yes, I've thought about max_share_supply too. But there is nothing that guarantees that an asset object will always contain an amount less than max_share_supply. I think the assets contained in price objects are particularly endangered here, because it is like that computations are performed on prices., and it is perfectly legal to (for example) specify a price feed with 64 bit precision.